### PR TITLE
[Messenger] Adding final routing key to delay queue name

### LINKED ISSUE #31241: [Messenger] SF 4.3 AMQP publish with delay invalid x-dead-letter-routing-key
**Symfony version(s) affected**: 4.3

**Description**  
When publish a message to a queue A with a routing key A`, if the message handling fails, the message will be retried 3 times by default and also a DelayStamp will be added to the envelope.
So because the DelayStamp was added, the message will be published using the publishWithDelay of the AmqpExt\Connection class.
That function will also call the setupDelay function, which will setup a delay queue, by default name `delay_queue_1000`. This delay queue will have the argument `x-dead-letter-routing-key` to the routing key of the message (A` in this case), so after the message is released from this queue it can be re-published with the correct routing key.

However, a problem happens if a message is published to lets say a queue B with routing key B`. If the handling of this message also fails, the DelayStamp will also be added to the envelope, so a the delay queue will again be setup. However, since by default the same queue name `delay_queue_1000` will be used, and that queue already had the argument `x-dead-letter-routing-key` set to the routing key A`, then the same argument can not be set to the routing key B`.

**How to reproduce**  
Have two messages published with different routing keys (may also need to be published to different transports). When running the consume command for the first message, the handling should fail so the message should be retried 3 times. Run the consume command for the second message, which handling should also fail. Notice that an error with the message ` Server channel error: 406, message: PRECONDITION_FAILED - inequivalent arg   
  'x-dead-letter-routing-key' for queue 'delay_queue_1000'` will appear in the console.

**Possible Solution**  
Have multiple delay queues depending on routing key.
OR
Do not associate the delay queue with a routing key and use stamps to handle the routing key (not sure if this is possible)

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31241
| License       | MIT
| Doc PR        | n/a

Fixes #31241.

When we delay, we create a queue whose `x-message-ttl` matches the delay length and `x-dead-letter-routing-key` matches the original routing key used for the message. However, before this PR, the original routing key was not part of that queue's name. The result is that if two messages were delayed by the same length, but with different routing keys, the second would try to "redeclare" the existing delay queue with a new `x-dead-letter-routing-key`, resulting in an error similar to:

> Server channel error: 406, message: PRECONDITION_FAILED - inequivalent arg 'x-dead-letter-routing-key' for queue 'delay_queue_1000' 

Integration test was improved to catch this.

Cheers!

### CODE REVIEW

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php [hunk 1]
```diff
@@ -286,9 +291,13 @@ private function createDelayQueue(int $delay, ?string $routingKey)
         return $queue;
     }
 
-    private function getRoutingKeyForDelay(int $delay): string
+    private function getRoutingKeyForDelay(int $delay, ?string $finalRoutingKey): string
     {
-        return str_replace('%delay%', $delay, $this->connectionOptions['delay']['routing_key_pattern']);
+        return str_replace(
+            ['%delay%', '%routing_key%'],
+            [$delay, $finalRoutingKey ?: 'no_routing_key'],
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: This breaks again if somebody uses `no_routing_key` as routing key. I know it's unlikely but solvable by using an empty string as default.

File: src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpExtIntegrationTest.php [hunk 2]
```diff
@@ -206,4 +221,25 @@ private function createSerializer(): SerializerInterface
             new SerializerComponent\Serializer([new ObjectNormalizer(), new ArrayDenormalizer()], ['json' => new JsonEncoder()])
         );
     }
+
+    private function assertApproximateDuration($startTime, int $expectedDuration)
+    {
+        $actualDuration = microtime(true) - $startTime;
+
+        $this->assertEquals($expectedDuration, $actualDuration, 'Duration was not within expected range', .5);
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpExtIntegrationTest.php]: The delta argument is deprecated in phpunit 8. Maybe we should check if `assertEqualsWithDelta` exist as use this instead. See https://github.com/sebastianbergmann/phpunit/issues/3341

**Comment by fabpot** [comment]
Thank you @weaverryan.

File: src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpExtIntegrationTest.php [hunk 3]
```diff
@@ -206,4 +221,29 @@ private function createSerializer(): SerializerInterface
             new SerializerComponent\Serializer([new ObjectNormalizer(), new ArrayDenormalizer()], ['json' => new JsonEncoder()])
         );
     }
+
+    private function assertApproximateDuration($startTime, int $expectedDuration)
+    {
+        $actualDuration = microtime(true) - $startTime;
+
+        if (\method_exists([$this, 'assertEqualsWithDelta'])) {
+            $this->assertEqualsWithDelta($expectedDuration, $actualDuration, 'Duration was not within expected range', .5);
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpExtIntegrationTest.php]: the order of arguments is wrong. `assertEqualsWithDelta($expected, $actual, float $delta, string $message = '')`

Comment (weaverryan) [reply src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpExtIntegrationTest.php]: I'm on it

**Pull request merged**